### PR TITLE
Fix Kafka partition key callback, when using synchronous wrapper

### PR DESCRIPTION
**Pull request by SpeedyCoder**
Currently when using synchronous adapter and Kafka we can't easily retrieve partition key from the substrate message, as the callback is called with an annotated one. With this change the user will receive the message they originally provided. We can potentially make the `AnnotatedMessage` interface private, but it will make implementing additional sink wrappers outside of the package easier.

### CODE REVIEW

File: substrate.go [hunk 1]
```diff
@@ -23,6 +23,15 @@ type DiscardableMessage interface {
 	DiscardPayload()
 }
 
+// AnnotatedMessage is an interface implemented by messages used by sink wrappers,
+// in particular the synchronous sink adapter. This allows the backends to retrieve
+// the original message the user sent when invoking user provided callbacks for example
+// for getting partition keys.
+type AnnotatedMessage interface {
+	Message
+	OriginalMessage() Message
```
Comment (davidsbond) [inline substrate.go]: IMO I'd change the method name to `Original`, I think have `Message` as a suffix when it returns type `Message` is a little repetitive.

**Comment by mjgarton** [comment]
This all looks nice and generic, but it really just makes it easier to bypass the substrate abstractions and make your code become broker specific.  How will code that's unwrapping to get at kafka specifics remain portable?

**Comment by SpeedyCoder** [comment]
@mjgarton The Kafka specific code will obviously never become generic. This doesn't change that, the stuff that was Kafka specific will remain so. This PR just makes the synchronous wrapper easily usable with partitioned Kafka, what's not the case at the moment. It also adds a mechanism for unwrapping messages to get the user provided one in actual backed layer, so it can invoke user provided callback with the message that the user originally provided. The only use case for this mechanism at the moment is the partition key retrieving, but it can be used in many more scenarios.

**Comment by SpeedyCoder** [comment]
@mjgarton I've moved the annotated message interface to the internal unwrap package as we've discussed, so this should be good to go.

**Pull request merged**